### PR TITLE
Persist log messages across app restarts with Core Data

### PR DESCRIPTION
**Pull request by JesseACooper**
Super interested in any style or function ergonomics feedback people have - still not super comfortable with Core Data (or Swift, for that matter).

### CODE REVIEW

File: root-drive-science-ios/LogMessage+CoreDataProperties.swift [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+import CoreData
+import Foundation
+
+extension LogMessage {
+    @nonobjc public class func fetchRequest() -> NSFetchRequest<LogMessage> {
+        return NSFetchRequest<LogMessage>(entityName: "LogMessage")
+    }
+
+    @NSManaged public var date: Date?
```
Comment (frankmassi-root) [inline root-drive-science-ios/LogMessage+CoreDataProperties.swift:18]: Are we not storing these as full timestamps, or is this just swift not being clear?

File: root-drive-science-ios/ViewController.swift [hunk 2]
```diff
@@ -102,13 +125,42 @@ class ViewController: UIViewController {
...
-    func appendNotificationText(_ text: String) {
+    private func saveLogMessage(_ text: String) -> LogMessage? {
+        let managedContext = container.viewContext
+
+        let entity = NSEntityDescription.entity(forEntityName: "LogMessage", in: managedContext)!
+
+        let logMessage = LogMessage(entity: entity, insertInto: managedContext)
+
+        logMessage.message = text
+        logMessage.date = Date()
```
Comment (frankmassi-root) [inline root-drive-science-ios/ViewController.swift:146]: Same line of questioning...should this be a full timestamp?

**Review by frankmassi-root — COMMENTED** [review]
Just one question about the datatype we're storing for the timestamp of the log message.

File: root-drive-science-ios/ViewController.swift [hunk 3]
```diff
@@ -102,13 +125,42 @@ class ViewController: UIViewController {
...
+            print("Could not save log message. \(error), \(error.userInfo)")
+        }
+
+        return nil
+    }
+
+    private func displayLogMessage(_ logMessage: LogMessage) {
         DispatchQueue.main.async {
-            self.notificationField.text = self.notificationField.text.appending("[\(self.timestamp)]:" + text + "\n")
+            let message = logMessage.message!
```
Comment (atreat) [inline root-drive-science-ios/ViewController.swift:161]: The force unwrap here is dangerous and will cause crashes if logMessage ever has a nil message or date. Would recommend wrapping this in an if/let

```
if let message = logMessage.message, let date = logMessage.date {
    self.notificationField.text = // ...
}
```

It is also possible to make these fields required in the data model metadata and then the properties would not be optional.

File: root-drive-science-ios/LogMessage+CoreDataProperties.swift (hunk 1 above)
Comment (JesseACooper) [reply root-drive-science-ios/LogMessage+CoreDataProperties.swift:18]: These do include time, so in the app logs they'll show up as e.g. `[2020-09-25 16:30]`

File: root-drive-science-ios/ViewController.swift (hunk 3 above)
Comment (JesseACooper) [reply root-drive-science-ios/ViewController.swift:161]: 👍

**Review by erinnachen — APPROVED** [review]

**Review by atreat — APPROVED** [review]

**Pull request merged**